### PR TITLE
Switch to lodash

### PR DESCRIPTION
**Pull request by aweidner**
Changed everything over from underscore to lodash and also refactored a bit.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,28 +8,30 @@
   },
   "dependencies": {
     "async": "2.0.0-rc.3",
+    "async-lock": "0.3.8",
```
Comment (endrawes0) [inline package.json]: Just realized this was alphabetized... well done.

File: src/chesster.js [hunk 2]
```diff
@@ -422,8 +422,7 @@ chesster.controller.hears([
 	'direct_mention', 
 	'direct_message'
 ], function(bot,message) {
-    bot_exception_handler(bot, message, function(){
-    });
+    bot_exception_handler(bot, message, _.noop)
```
Comment (endrawes0) [inline src/chesster.js]: oh i like this. such simply things that I tend to do for myself. I never thought to look for them in a utility.

File: src/league.js [hunk 3]
```diff
@@ -460,11 +460,7 @@ league_attributes = {
     'getCaptains':function() {
         var self = this;
         return Q.fcall(function() {
-            var captains = [];
-            _.each(self._teams, function(team) {
-                captains.push(team.captain);
-            });
-            return captains;
+            return _.map(self._teams, "captain");
```
Comment (endrawes0) [inline src/league.js]: so this syntax is a little weird to me.

I expect `map` to apply a function to every `team` in `_teams`. 

How should I be reading this?

Clearly it returns an array of captains from `_teams` but it doesnt do this by applying a function - atleast not at this level of abstraction.

File: src/league.js [hunk 4]
```diff
@@ -647,12 +643,13 @@ function League(options) {
 }
 
 function getAllLeagues(config) {
-    var leagues = [];
     var all_league_configs = config['leagues'] || {};
-    _.each(_.keys(all_league_configs), function(key) {
-        leagues.push(getLeague(key, config));
-    });
-    return leagues;
+    return _(all_league_configs)
```
Comment (endrawes0) [inline src/league.js]: I suppose I should get used to this syntax. I would not have written it this way. I tend to code in a more procedural style - all that C++98...

File: src/spreadsheets.js [hunk 5]
```diff
@@ -407,7 +407,10 @@ function getRows(spreadsheetConfig, options, sheetPredicate, callback) {
                     });
                     var recordRows = [];
                     rows.slice(2).forEach(function(row) {
-                        recordRows.push(_.object(_.zip(headerRow, row)));
+                        var asObject = _.transform(_.zip(headerRow, row), function(result, pair) {
```
Comment (endrawes0) [inline src/spreadsheets.js]: I have to learn about lodash. I love the work it does. I would have coded all this manually.

File: src/subscription.js [hunk 6]
```diff
@@ -262,11 +258,7 @@ function getListeners(leagueName, source, event) {
             }
         }).then(function(subscriptions) {
             unlock.resolve();
-            var listeners = [];
-            _.each(subscriptions, function(subscription) {
-                listeners.push(subscription.target);
-            });
-            return _.uniq(listeners);
+            return _(subscriptions).map("target").uniq().value();
```
Comment (endrawes0) [inline src/subscription.js]: I love how this reads like a unix pipe.

**Comment by endrawes0** [comment]
Looks great to me. Do as @lakinwecker suggests. I am digging lodash.

File: src/spreadsheets.js (hunk 5 above)
Comment (lakinwecker) [inline src/spreadsheets.js]: This is one area where I think I preferred the underscore version. :(

File: src/league.js (hunk 3 above)
Comment (aweidner) [inline src/league.js]: So this syntax is basically [team.captain for team in teams]

File: src/spreadsheets.js (hunk 5 above)
Comment (aweidner) [inline src/spreadsheets.js]: I agree, the underscore version would be much preferred here, unfortunately `_.object` is not defined in lodash and I wasn't able to find a suitable replacement.

File: src/league.js (hunk 3 above)
Comment (endrawes0) [inline src/league.js]: That makes sense. Thanks for the explanation.

File: src/spreadsheets.js (hunk 5 above)
Comment (lakinwecker) [inline src/spreadsheets.js]: I found this in lodash :heart: - ```var asObject =_.zipObject(headerRow, row);```

**Pull request merged**